### ISSUE TITLE
the mppt can get stuck in a loop

### ISSUE THREAD

**Issue by greyltc** [greyltc/solar-sim-software#4]
```
WARNING: Reached high voltage edge because we hit Voc
Reached low voltage edge because angle exceeded
WARNING: Reached high voltage edge because we hit Voc
Reached low voltage edge because angle exceeded
WARNING: Reached high voltage edge because we hit Voc
Reached low voltage edge because angle exceeded
WARNING: Reached high voltage edge because we hit Voc
Reached low voltage edge because angle exceeded
WARNING: Reached high voltage edge because we hit Voc
Reached low voltage edge because angle exceeded
WARNING: Reached high voltage edge because we hit Voc
Reached low voltage edge because angle exceeded
WARNING: Reached high voltage edge because we hit Voc
Reached low voltage edge because angle exceeded
WARNING: Reached high voltage edge because we hit Voc
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
WARNING: Reached high voltage edge because we hit Voc
Reached low voltage edge because angle exceeded
WARNING: Reached high voltage edge because we hit Voc
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
Reached low voltage edge because angle exceeded
```
Need to figure out how to reliably recreate this.
Maybe this can only happen with __disconnected pixels__ ~(or one of PIN vs NIP)~?
Still, it must be impossible under any circumstances.

**Comment by greyltc** [comment]
This apparently only happens with a disconnected pixel. ie. it's not responding as expected during the exploration phases.

**Comment by rbrenesh** [comment]
I ran into a similar problem while MPPT a device over 2 hours. However, it seems like the Keithley hit the compliance limit for the current. Attached is a picture of the command line when it happened.

However, the device was still operational after this with a much reduced Vmpp of 0.3V vs ~0.8V where it started. The device wasn't shorted as far as I could tell.

![Screenshot from 2019-03-13 14-43-42](https://user-images.githubusercontent.com/33428877/54322331-2e5c7680-45cb-11e9-9c6c-c87e30573243.png)

**Comment by greyltc** [comment]
First off, I'm surprised the Keithley hit the compliance limit during the MPPT. I set the compliance here to be 2x the value for the current at the maximum power point as found during a previous J-V sweep. I think probably the MPPT algorithm erroneously went beyond Voc and fully turned on the device's diode which is why the current went into compliance.

I've made changes that I hope will prevent endless loops while MPPTing. In 06853af7e8cfc9aff5f2693a627859620435b6a7 I've put some checks in that should prevent the basic MPPT algorithm from getting stuck behind "edges" that it should transition off of during the exploration phases.

Let's re-open this if we see any new endless loops.
